### PR TITLE
Allow to tweak default scopes for accounts

### LINKED ISSUE #6582: Tweak the default scopes for Accounts
Currently the following default scopes for new accounts are defined inside `AccountManager`. We should let the server admin define them through the administration panel or config.php.
This is especially important on instances where users don't know each other.
```php
self::PROPERTY_DISPLAYNAME =>
    [
        'value' => $user->getDisplayName(),
        'scope' => self::VISIBILITY_CONTACTS_ONLY,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_ADDRESS =>
    [
        'value' => '',
        'scope' => self::VISIBILITY_PRIVATE,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_WEBSITE =>
    [
        'value' => '',
        'scope' => self::VISIBILITY_PRIVATE,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_EMAIL =>
    [
        'value' => $user->getEMailAddress(),
        'scope' => self::VISIBILITY_CONTACTS_ONLY,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_AVATAR =>
    [
        'scope' => self::VISIBILITY_CONTACTS_ONLY
    ],
self::PROPERTY_PHONE =>
    [
        'value' => '',
        'scope' => self::VISIBILITY_PRIVATE,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_TWITTER =>
    [
        'value' => '',
        'scope' => self::VISIBILITY_PRIVATE,
        'verified' => self::NOT_VERIFIED,
    ],
```

### LINKED ISSUE #14358: set profile to private/public on new user creation
I'm always frustrated when I create multiple users and for them not to show up as contact in all other user contact lists, I have to log in as the new user and set the profile to private.

### Solution:

An option box to set the profile private in the user create dialogue

### Alternatives: 

1. set users profile to private per default, but this could lead to frustration if others want to have all users public per default
2. a setting in the config file that defines private or public for new users

### LINKED ISSUE #14959: don't reveal users mail adress by default
### Steps to reproduce
1. register as a new user
2. after some months, explore the menu
3. discover that your mail adress was public to ALL the other cloud users on the same server

### Expected behaviour
Private data should not be revealed **by default**
> Make public data available, protect private data.
Source: 
- hacker ethics
- gdpr

### Actual behaviour
Users mail adress is revealed to other users by default

### Server configuration

LAMP, Ubuntu 18.04

**Nextcloud version:** 15.0.5
fresh installl since upgrade from OC failed

**Where did you install Nextcloud from:** Webclient

**List of activated apps:**
deactivated ALL Apps - same behavior

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
If you have access to your command line run e.g.:

```{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.REMOVED.de"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "15.0.5.3",
        "overwrite.cli.url": "https:\/\/cloud.REMOVED.de",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "pipe",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "updater.secret": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "theme": "",
        "loglevel": 2,
        "CUSTOMSETTINGSMATTHIAS": "ab hier!!!",
        "default_language": "de",
        "default_locale": "de",
        "force_language": "de",
        "force_locale": "de",
        "mail_smtpsecure": "ssl",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***"
    }
}```

### Client configuration
**Browser:**
IE, FF, Chrome
**Operating system:**

### PR DESCRIPTION
**Pull request by tcitworld**
Closes #6582, Closes #14358, Closes #14959

Lacks proper tests

Needs to be documented afterwards.

### CODE REVIEW

File: lib/private/Accounts/AccountManager.php [hunk 1]
```diff
@@ -298,50 +307,58 @@ protected function updateExistingUser(IUser $user, $data) {
 	 * @return array
 	 */
 	protected function buildDefaultUserRecord(IUser $user) {
+		$override = $this->config->getSystemValue('account_manager_default_property_scope', []);
 		return [
 			self::PROPERTY_DISPLAYNAME =>
 				[
 					'value' => $user->getDisplayName(),
-					'scope' => self::VISIBILITY_CONTACTS_ONLY,
+					'scope' => $this->getOverrideValueOrDefault($override,self::PROPERTY_DISPLAYNAME),
```
Comment (tcitworld) [inline lib/private/Accounts/AccountManager.php:315]: CS doesn't catch this missing space?

File: lib/private/Accounts/AccountManager.php [hunk 2]
```diff
@@ -298,50 +308,58 @@ protected function updateExistingUser(IUser $user, $data) {
 	 * @return array
 	 */
 	protected function buildDefaultUserRecord(IUser $user) {
+		$override = $this->config->getSystemValue('account_manager_default_property_scope', []);
```
Comment (kesselb) [inline lib/private/Accounts/AccountManager.php:311]:
```
		$scopes = array_merge(self::DEFAULT_SCOPE_VALUES, array_filter($this->config->getSystemValue('account_manager_default_property_scope', []), function ($scope) {
			return in_array($scope, self::VISIBILITIES_LIST, true);
		}));
```

I think you can simplify `getOverrideValueOrDefault` a bit with array_merge if you want.

File: config/config.sample.php [hunk 3]
```diff
@@ -1805,4 +1805,11 @@
  */
 
 'login_form_autocomplete' => true,
+
+/**
+ * Allows to override the default visibility scopes for Account data.
+ * The list of properties overridable and the visibilities possible values are in
```
Comment (kesselb) [inline config/config.sample.php:1811]:
```suggestion
 * The list of overridable properties and valid values for visibility are in
```

**Review by kesselb — COMMENTED** [review]
Good stuff :+1:

File: lib/private/Accounts/AccountManager.php (hunk 2 above)
Comment (tcitworld) [reply lib/private/Accounts/AccountManager.php]: Much better, thank you very much!

File: lib/private/Accounts/AccountManager.php [hunk 4]
```diff
@@ -67,21 +69,29 @@ class AccountManager implements IAccountManager {
...
-	 *
-	 * @param IDBConnection $connection
-	 * @param EventDispatcherInterface $eventDispatcher
-	 * @param IJobList $jobList
-	 */
+	/** @var IConfig */
+	private $config;
+
+	public const DEFAULT_SCOPE_VALUES = [
+		self::PROPERTY_DISPLAYNAME => self::VISIBILITY_CONTACTS_ONLY,
```
Comment (nickvergessen) [inline lib/private/Accounts/AccountManager.php:76]: We should really make the default on display name and avatar public.
Talk and some other apps are completly unusable at the moment when the settings would be respected.

File: lib/private/Accounts/AccountManager.php [hunk 5]
```diff
@@ -67,21 +69,29 @@ class AccountManager implements IAccountManager {
...
-	 * AccountManager constructor.
-	 *
-	 * @param IDBConnection $connection
-	 * @param EventDispatcherInterface $eventDispatcher
-	 * @param IJobList $jobList
-	 */
+	/** @var IConfig */
+	private $config;
+
+	public const DEFAULT_SCOPE_VALUES = [
```
Comment (ChristophWurst) [inline lib/private/Accounts/AccountManager.php:75]:
```suggestion
	public const DEFAULT_SCOPES = [
```

File: lib/public/Accounts/IAccountManager.php [hunk 6]
```diff
@@ -52,6 +59,16 @@ interface IAccountManager {
 	public const PROPERTY_ADDRESS = 'address';
 	public const PROPERTY_TWITTER = 'twitter';
 
+	public const PROPERTIES_LIST = [
```
Comment (ChristophWurst) [inline lib/public/Accounts/IAccountManager.php:62]:
```suggestion
	public const PROPERTIES = [
```

**Review by ChristophWurst — COMMENTED** [review]
Nitpicks on naming

File: lib/private/Accounts/AccountManager.php (hunk 4 above)
Comment (tcitworld) [reply lib/private/Accounts/AccountManager.php:76]: Do you want to have the change in this PR or shall I open an issue to discuss instead?

**Comment by kesselb** [comment]
@tcitworld 👋

**Comment by nickvergessen** [comment]
Btw the meaning (and labels) were changed in the meantime and they are now not about visibility but whether they are synced to trusted server and the public addressbook/lookup server.

So maybe visibility is a wrong constant name nowadays

**Comment by Dubidubiduu** [comment]
still not fixed, right?

**Comment by J0WI** [comment]
ping @tcitworld

**Comment by MorrisJobke** [comment]
Let's close this for now. We can reopen it later at any time.

**Comment by tcitworld** [comment]
My branch though :(

**Comment by MorrisJobke** [comment]
> My branch though :(

It can be restored from the github UI at any time :)

**Comment by tcitworld** [comment]
Follow-up at #31623

**Pull request closed without merging**